Latest main branch breaks testbed `test_app.py::test_current_window` test on macOS

### Describe the bug

With the latest main branch of toga, at the beginning of every test of testbed, an "Open File Dialog" is being shown, which has input focus. Consequently the testbed `test_app.py::test_current_window` test fails as the input focus is on the dialog instead of on the main window. 

### Steps to reproduce

1. Clone the latest toga main branch
```
git clone https://github.com/beeware/toga.git
cd toga/testbed
```
2. Run the testbed test_app.py::test_current_window:
```
briefcase dev -r  --test -- tests/app/test_desktop.py::test_current_window -s
```
3. Notice a "Open File dialog" being shown from the start of the test:
![image](https://github.com/user-attachments/assets/756a1f10-2c5f-4aee-9522-140b4177d0a0)

4. Get Error Message:
```
[testbed] Running test suite in dev environment...
===========================================================================
Waiting for app to be ready for testing... ready.
============================= test session starts ==============================
collecting ... collected 1 item

tests/app/test_desktop.py::test_current_window FAILED

=================================== FAILURES ===================================
_____________________________ test_current_window ______________________________
Traceback (most recent call last):
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
    yield from thread_exception_runtest_hook()
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/logging.py", line 846, in pytest_runtest_call
    yield from self._runtest_for(item, "call")
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/logging.py", line 829, in _runtest_for
    yield
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/capture.py", line 880, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
    item.runtest()
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pytest_asyncio/plugin.py", line 457, in runtest
    super().runtest()
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/python.py", line 1627, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/pytest_asyncio/plugin.py", line 929, in inner
    _loop.run_until_complete(task)
  File "/Users/proneon267/toga/testbed/tests/conftest.py", line 142, in run_until_complete
    return asyncio.run_coroutine_threadsafe(coro, self.loop).result()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/.pyenv/versions/3.12.4/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/proneon267/.pyenv/versions/3.12.4/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/proneon267/toga/testbed/tests/app/test_desktop.py", line 353, in test_current_window
    assert app.current_window == main_window
  File "/Users/proneon267/venv/lib/python3.12/site-packages/toga/app.py", line 831, in current_window
    return window.interface
           ^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/rubicon/objc/api.py", line 1097, in __getattr__
    raise AttributeError(
AttributeError: rubicon.objc.api.ObjCInstance NSOpenPanel has no attribute interface
=========================== short test summary info ============================
FAILED tests/app/test_desktop.py::test_current_window - AttributeError: rubicon.objc.api.ObjCInstance NSOpenPanel has no attribute interface
============================== 1 failed in 0.63s ===============================
```

### Expected behavior

The "Open File Dialog" should not be shown at the start of the test.

### Screenshots

_No response_

### Environment

- Operating System: macOS 14.6
- Python version: 3.12
- Software versions:
  - Briefcase: 0.3.19
  - Toga: latest main branch
  - ...


### Logs

```
No logs generated.
```


### Additional context

The bugs seems to have been arisen out of the following addition of the document_types to the testbed:
https://github.com/beeware/toga/blob/8bec404dd732d478b3ca50b1ccfefcebf5c0e650/testbed/src/testbed/app.py#L225-L229
As without specifying the document_types, the testbed `test_app.py::test_current_window` test passes and no "Open File Dialog" is shown at the beginning of the test.

This problem is also arising on the linux-wayland testbed, as evident by the CI failures of input focus related widget tests.

--- a/changes/2860.bugfix.rst
+++ b/changes/2860.bugfix.rst
@@ -1,1 +1,1 @@
-Apps that specify both `document_types` and a `main_window` no longer display the "Document Selection" dialog on startup in macOS.
+On macOS, apps that specify both `document_types` and a `main_window` no longer display the document selection dialog on startup.

--- a/changes/2860.bugfix.rst
+++ b/changes/2860.bugfix.rst
@@ -1,0 +1,1 @@
+Apps that specify both `document_types` and a `main_window` no longer display the "Document Selection" dialog on startup in macOS.

--- a/cocoa/src/toga_cocoa/app.py
+++ b/cocoa/src/toga_cocoa/app.py
@@ -54,7 +54,10 @@ class AppDelegate(NSObject):
 
     @objc_method
     def applicationShouldOpenUntitledFile_(self, sender) -> bool:
-        return bool(self.interface.documents.types)
+        if self.interface._main_window == toga.App._UNDEFINED:
+            return False
+        else:
+            return bool(self.interface.documents.types)
 
     @objc_method
     def application_openFiles_(self, app, filenames) -> None:

--- a/core/src/toga/app.py
+++ b/core/src/toga/app.py
@@ -608,35 +608,34 @@ class App:
 
     def _create_initial_windows(self):
         """Internal utility method for creating initial windows based on command line
-        arguments. This method is used when the platform doesn't provide its own
-        command-line handling interface.
+        arguments.
 
-        If document types are defined, try to open every argument on the command line as
-        a document (unless the backend manages the command line arguments).
+        If document types are defined, and the backend doesn't have native command line
+        handling, try to open every argument on the command line as a document (unless
+        the backend manages the command line arguments).
+
+        If, after processing all command line arguments, the app doesn't have at least
+        one window, the app's default initial document handling will be triggered.
         """
-
-        if not (self._impl.HANDLES_COMMAND_LINE and self._impl.CLOSE_ON_LAST_WINDOW):
-            doc_count = len(self.windows)
-
-            # Process command line arguments if the backend doesn't handle them
-            if not self._impl.HANDLES_COMMAND_LINE and self.documents.types:
+        # Process command line arguments if the backend doesn't handle them
+        if not self._impl.HANDLES_COMMAND_LINE:
+            if self.documents.types:
                 for filename in sys.argv[1:]:
-                    if self._open_initial_document(filename):
-                        doc_count += 1
+                    self._open_initial_document(filename)
 
-            # Ensure there is at least one document or window
-            if self.main_window is None and doc_count == 0:
-                if self.documents.types:
-                    if self._impl.CLOSE_ON_LAST_WINDOW:
-                        # Pass in the first document type as the default
-                        self.documents.new(self.documents.types[0])
-                    else:
-                        self.loop.run_until_complete(self.documents.request_open())
+        # Ensure there is at least one window
+        if self.main_window is None and len(self.windows) == 0:
+            if self.documents.types:
+                if self._impl.CLOSE_ON_LAST_WINDOW:
+                    # Pass in the first document type as the default
+                    self.documents.new(self.documents.types[0])
                 else:
-                    # No document types defined.
-                    raise RuntimeError(
-                        "App didn't create any windows, or register any document types."
-                    )
+                    self.loop.run_until_complete(self.documents.request_open())
+            else:
+                # No document types defined.
+                raise RuntimeError(
+                    "App didn't create any windows, or register any document types."
+                )
 
     def _startup(self) -> None:
         # Wrap the platform's event loop's task factory for task tracking

--- a/core/src/toga/app.py
+++ b/core/src/toga/app.py
@@ -493,14 +493,14 @@ class App:
 
         def factory(loop, coro, context=None):
             if platform_task_factory is not None:
-                if sys.version_info < (3, 11):
+                if sys.version_info < (3, 11):  # pragma: no-cover-if-gte-py311
                     task = platform_task_factory(loop, coro)
-                else:
+                else:  # pragma: no-cover-if-lt-py311
                     task = platform_task_factory(loop, coro, context=context)
             else:
-                if sys.version_info < (3, 11):
+                if sys.version_info < (3, 11):  # pragma: no-cover-if-gte-py311
                     task = asyncio.Task(coro, loop=loop)
-                else:
+                else:  # pragma: no-cover-if-lt-py311
                     task = asyncio.Task(coro, loop=loop, context=context)
 
             self._running_tasks.add(task)

--- a/core/src/toga/app.py
+++ b/core/src/toga/app.py
@@ -659,6 +659,15 @@ class App:
         # Create any initial windows
         self._create_initial_windows()
 
+        # If the app has document types but has no windows, then show an "Open Document"
+        # Dialog for the user to select a document.
+        if (
+            self.main_window is None
+            and len(self.windows) == 0
+            and self.documents.types != []
+        ):
+            asyncio.create_task(self.documents.request_open())
+
         # Manifest the initial state of the menus. This will cascade down to all
         # open windows if the platform has window-based menus. Then install the
         # on-change handler for menus to respond to any future changes.

--- a/core/src/toga/app.py
+++ b/core/src/toga/app.py
@@ -614,26 +614,29 @@ class App:
         If document types are defined, try to open every argument on the command line as
         a document (unless the backend manages the command line arguments).
         """
-        # If the backend handles the command line, don't do any command line processing.
-        if self._impl.HANDLES_COMMAND_LINE:
-            return
-        doc_count = len(self.windows)
-        if self.documents.types:
-            for filename in sys.argv[1:]:
-                if self._open_initial_document(filename):
-                    doc_count += 1
 
-        # Safety check: Do we have at least one document?
-        if self.main_window is None and doc_count == 0:
-            try:
-                # Pass in the first document type as the default
-                default_doc_type = self.documents.types[0]
-                self.documents.new(default_doc_type)
-            except IndexError:
-                # No document types defined.
-                raise RuntimeError(
-                    "App didn't create any windows, or register any document types."
-                )
+        if not (self._impl.HANDLES_COMMAND_LINE and self._impl.CLOSE_ON_LAST_WINDOW):
+            doc_count = len(self.windows)
+
+            # Process command line arguments if the backend doesn't handle them
+            if not self._impl.HANDLES_COMMAND_LINE and self.documents.types:
+                for filename in sys.argv[1:]:
+                    if self._open_initial_document(filename):
+                        doc_count += 1
+
+            # Ensure there is at least one document or window
+            if self.main_window is None and doc_count == 0:
+                if self.documents.types:
+                    if self._impl.CLOSE_ON_LAST_WINDOW:
+                        # Pass in the first document type as the default
+                        self.documents.new(self.documents.types[0])
+                    else:
+                        self.loop.run_until_complete(self.documents.request_open())
+                else:
+                    # No document types defined.
+                    raise RuntimeError(
+                        "App didn't create any windows, or register any document types."
+                    )
 
     def _startup(self) -> None:
         # Wrap the platform's event loop's task factory for task tracking
@@ -658,15 +661,6 @@ class App:
 
         # Create any initial windows
         self._create_initial_windows()
-
-        # If the app has document types but has no windows, then show an "Open Document"
-        # Dialog for the user to select a document.
-        if (
-            self.main_window is None
-            and len(self.windows) == 0
-            and self.documents.types != []
-        ):
-            asyncio.create_task(self.documents.request_open())
 
         # Manifest the initial state of the menus. This will cascade down to all
         # open windows if the platform has window-based menus. Then install the

--- a/core/tests/app/test_document_app.py
+++ b/core/tests/app/test_document_app.py
@@ -1,5 +1,5 @@
 import sys
-from unittest.mock import Mock
+from unittest.mock import AsyncMock, MagicMock, Mock
 
 import pytest
 
@@ -151,7 +151,7 @@ def test_create_no_cmdline(monkeypatch):
     assert toga.Command.SAVE_ALL in app.commands
 
 
-def test_create_no_cmdline_default_handling(monkeypatch):
+def test_create_no_cmdline_default_handling_close_on_last_window(monkeypatch):
     """If the backend uses the app's command line handling, no error is raised for an
     empty command line."""
     monkeypatch.setattr(sys, "argv", ["app-exe"])
@@ -173,6 +173,36 @@ def test_create_no_cmdline_default_handling(monkeypatch):
     # No documents or windows exist
     assert len(app.documents) == 0
     assert len(app.windows) == 0
+
+
+def test_create_no_cmdline_default_handling_no_close_on_last_window(monkeypatch):
+    """If the backend handles the app's command line and the app doesn't exit when
+    the last window closes, no error is raised for an empty command line and the
+    app shows a document selection dialog on startup."""
+
+    monkeypatch.setattr(sys, "argv", ["app-exe"])
+
+    # Monkeypatch the property that makes the backend handle command line arguments
+    # and not close the app when the last window closes.
+    monkeypatch.setattr(DummyApp, "HANDLES_COMMAND_LINE", True)
+    monkeypatch.setattr(DummyApp, "CLOSE_ON_LAST_WINDOW", False)
+
+    # Mock request_open() because OpenFileDialog's response can't be set before the
+    # app creation. Since request_open() is called during app initialization, we can't
+    # set the dialog response in time, leading to an unexpected dialog response error.
+    monkeypatch.setattr(ExampleDocumentApp, "documents", MagicMock())
+    mock_request_open = AsyncMock()
+    monkeypatch.setattr(ExampleDocumentApp.documents, "request_open", mock_request_open)
+
+    # Create the app instance
+    _ = ExampleDocumentApp(
+        "Test App",
+        "org.beeware.document-app",
+        document_types=[ExampleDocument],
+    )
+
+    # Check that request_open was called
+    mock_request_open.assert_called_once()
 
 
 def test_create_with_cmdline(monkeypatch, example_file):

--- a/core/tests/app/test_document_app.py
+++ b/core/tests/app/test_document_app.py
@@ -1,9 +1,10 @@
 import sys
-from unittest.mock import AsyncMock, MagicMock, Mock
+from unittest.mock import AsyncMock, Mock
 
 import pytest
 
 import toga
+from toga.documents import DocumentSet
 from toga_dummy.app import App as DummyApp
 from toga_dummy.command import Command as DummyCommand
 from toga_dummy.utils import (
@@ -170,9 +171,9 @@ def test_create_no_cmdline_default_handling_close_on_last_window(monkeypatch):
 
     assert app.documents.types == [ExampleDocument]
 
-    # No documents or windows exist
-    assert len(app.documents) == 0
-    assert len(app.windows) == 0
+    # A default document has been created
+    assert len(app.documents) == 1
+    assert len(app.windows) == 1
 
 
 def test_create_no_cmdline_default_handling_no_close_on_last_window(monkeypatch):
@@ -190,18 +191,26 @@ def test_create_no_cmdline_default_handling_no_close_on_last_window(monkeypatch)
     # Mock request_open() because OpenFileDialog's response can't be set before the
     # app creation. Since request_open() is called during app initialization, we can't
     # set the dialog response in time, leading to an unexpected dialog response error.
-    monkeypatch.setattr(ExampleDocumentApp, "documents", MagicMock())
     mock_request_open = AsyncMock()
-    monkeypatch.setattr(ExampleDocumentApp.documents, "request_open", mock_request_open)
+    monkeypatch.setattr(DocumentSet, "request_open", mock_request_open)
 
     # Create the app instance
-    _ = ExampleDocumentApp(
+    app = ExampleDocumentApp(
         "Test App",
         "org.beeware.document-app",
         document_types=[ExampleDocument],
     )
 
-    # Check that request_open was called
+    assert app._impl.interface == app
+    assert_action_performed(app, "create App")
+
+    assert app.documents.types == [ExampleDocument]
+
+    # No documents have been created
+    assert len(app.documents) == 0
+    assert len(app.windows) == 0
+
+    # ...but request_open was called
     mock_request_open.assert_called_once()
 
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -38,6 +38,8 @@ exclude_lines = [
 no-cover-if-missing-setuptools_scm = "not is_installed('setuptools_scm')"
 no-cover-if-missing-PIL = "not is_installed('PIL')"
 no-cover-if-PIL-installed = "is_installed('PIL')"
+no-cover-if-lt-py311 = "sys_version_info < (3, 11) and os_environ.get('COVERAGE_EXCLUDE_PYTHON_VERSION') != 'disable'"
+no-cover-if-gte-py311 = "sys_version_info > (3, 11) and os_environ.get('COVERAGE_EXCLUDE_PYTHON_VERSION') != 'disable'"
 no-cover-if-lt-py310 = "sys_version_info < (3, 10) and os_environ.get('COVERAGE_EXCLUDE_PYTHON_VERSION') != 'disable'"
 no-cover-if-gte-py310 = "sys_version_info > (3, 10) and os_environ.get('COVERAGE_EXCLUDE_PYTHON_VERSION') != 'disable'"
 no-cover-if-lt-py39 = "sys_version_info < (3, 9) and os_environ.get('COVERAGE_EXCLUDE_PYTHON_VERSION') != 'disable'"

